File: 0138-copy-list-with-random-pointer/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*,Node*> m; // hashmap for references
        m[NULL] = NULL; // for edge case, null points to null

        Node* curr = head;
        while (curr != NULL) {
            Node* copy = new Node(curr->val);
            m[curr] = copy;
            curr = curr->next;
        }

        curr = head;
        while (curr != NULL) {
            Node* copy = m[curr];
            copy->next = m[curr->next];
            copy->random = m[curr->random];
            curr = curr->next;
        }

        return m[head];
    }
};

File: 0138-copy-list-with-random-pointer/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        m = { None : None } # edge case, null -> null

        curr = head

        # PASS 1: Hashmap.
        while curr:
            copy = Node(curr.val)
            m[curr] = copy
            curr = curr.next
        
        curr = head

        # PASS 2: Create new linked list.
        while curr:
            copy = m[curr]
            copy.next = m[curr.next]
            copy.random = m[curr.random]
            curr = curr.next
        
        return m[head]
